### PR TITLE
Introducing statefulness for single process dataloader

### PR DESCRIPTION
**Pull request by ramanishsingh**
We (@scotts @aelavender @psaikaushik @divyanshk) are the maintainers of the torchdata (https://github.com/meta-pytorch/data)  repo.

We plan to upstream the stateful functionality of StatefulDataLoader (https://github.com/meta-pytorch/data/blob/main/torchdata/stateful_dataloader/stateful_dataloader.py) to the PyTorch DataLoader.

This PR is the first of two PRs in which we upstream statefulness functionality. This PR focuses on implementing statefulness for single process DataLoader and the next one will be for multiprocess DataLoader.

We first made sure that statefulness does not impact DataLoader speed

**Speed comparison between StatefulDataLoader and torch.utils.data DataLoader**
plot: speed comparison for num_workers=0, error bars denote standard deviation.
speedup_by_dataset_sp

plot: speed comparison for num_workers=4
speedup_by_dataset_mp

Script used for benchmarking  : https://gist.github.com/ramanishsingh/479b35d8733cc6035a29e0e16c728ff5
Plotting script (requires result files benchmark_results_nw0.json and benchmark_results_nw4.json): https://gist.github.com/ramanishsingh/05190a9b18a51f1cf42906ba760c2b42
We compared the speeds for different types of datasets. 

​​SyntheticDataset: Simulates image-like data
LargeDataset: Stresses memory/bandwidth with large tensors
SmallDataset: Stresses overhead with very small tensors
ComputeHeavyDataset: Simulates on-the-fly computation per sample
SyntheticIterableDataset: For streaming/iterable scenarios
ShuffledStatefulIterableDataset: Tests stateful shuffling and buffer management ( state dict contains 500 samples in the buffer)

We chose a batch size (different for each dataset) such that each batch takes around 0.2-1 s and simulated 64 batches for each experiment and conducted 16 independent experiments.

Once we made sure we do not see a drop in performance, we upstreamed statefulness to DataLoader for single process DataLoader (this PR) and compared the performance b/w StatefulDataLoader, DataLoader(.., stateful=True), and DataLoader(..., stateful=False) for the same Datasets.

**Speed comparison after upstreaming statefulness to torch.utils.data DataLoader**
Benchmark script: https://gist.github.com/ramanishsingh/4b2085257948681485bfd100a6acef31
Plotting script: https://gist.github.com/ramanishsingh/a08943b7e55674f245df1302a8acc15e
plot: speed comparison for num_workers=0 forStatefulDataLoader, DataLoader(.., stateful=True), and DataLoader(..., stateful=False).
dataloader_speedup_comparison_nw0

Here as well, we do not see statistical difference between the speeds of the three settings. 
Thus, we would like to upstream the statefulness functionality of StatefulDataLoader to DataLoader, starting with the single process dataloader. 

Please note that here we are just focusing on the cost of maintaining state. 
The actual operation of getting the state_dict (`dataloader.state_dict()`) and loading it (`dataloader.load_state_dict()`) can be much more expensive, depends highly on the type of  `Dataset` used (for example in `ShuffledStatefulIterableDataset` we keep a buffer of actual elements which can be expensive to get, save, and reload), and is mostly saved for every 100 or 1000 steps. 

cc @andrewkho @divyanshk @SsnL @VitalyFedyunin @dzhulgakov @scotts

### CODE REVIEW

**Review by divyanshk — COMMENTED** [review]
@ramanishsingh can you import this PR internally and run unit tests?
